### ISSUE TITLE
RL runs so slowly

### ISSUE THREAD

**Issue by Pang-0093** [lqtrung1998/mwp_ReFT#28]
Hi, I find that the RL stage runs so slowly. Each epoch consumes nearly two hours, given the bs=1 and 8 A100. The default setting in the paper suggests running RL for 300 epochs, which means that we have to complete the RL stage with almost 1 month 😢. Is there anything wrong with my situation? Or is it normal? If it is normal, can we do anything to accelerate the RL stage's training?

**Comment by aivolcano** [comment]
Same problem, can we reduce the number of training data? I remember that OpenAI mentioned only about 100 records to train the model in rl stage for this method

**Comment by aivolcano** [comment]
> Hi, I find that the RL stage runs so slowly. Each epoch consumes nearly two hours, given the bs=1 and 8 A100. The default setting in the paper suggests running RL for 300 epochs, which means that we have to complete the RL stage with almost 1 month 😢. Is there anything wrong with my situation? Or is it normal? If it is normal, can we do anything to accelerate the RL stage's training?

The finial checkpoint in SFT stage is epoch=2. an issue mentioned epoch=2 is accepted. Do we must run 300 epochs in rl stage.

**Comment by aivolcano** [comment]
> Hi, I find that the RL stage runs so slowly. Each epoch consumes nearly two hours, given the bs=1 and 8 A100. The default setting in the paper suggests running RL for 300 epochs, which means that we have to complete the RL stage with almost 1 month 😢. Is there anything wrong with my situation? Or is it normal? If it is normal, can we do anything to accelerate the RL stage's training?

@Pang-0093

**Comment by Pang-0093** [comment]
> > Hi, I find that the RL stage runs so slowly. Each epoch consumes nearly two hours, given the bs=1 and 8 A100. The default setting in the paper suggests running RL for 300 epochs, which means that we have to complete the RL stage with almost 1 month 😢. Is there anything wrong with my situation? Or is it normal? If it is normal, can we do anything to accelerate the RL stage's training?
> 
> The finial checkpoint in SFT stage is epoch=2. an issue mentioned epoch=2 is accepted. Do we must run 300 epochs in rl stage.

As far as I have observed, 2 epochs for the RL stage is obviously not enough. Take 'python engine, codellama, gsm8k' setting as an example, 2 epochs' RL only leads to ~65%. However, I got 71.3% at 15th epoch and it remains the same during the following 10 epochs. (that's all I have run so far.😂)

**Comment by aivolcano** [comment]
Thank u for your updates. Are you test pcot format data set. The final metric score is about 75.28, which outperforms SFT and online self training a lot. If we take the computational efficiency into account. I think 15 epochs is enough. Hahaha

**Comment by aivolcano** [comment]
![image](https://user-images.githubusercontent.com/68730894/400296974-ba263d10-b86e-4889-8c0a-c997f67e25aa.png)

**Comment by aivolcano** [comment]
![image](https://user-images.githubusercontent.com/68730894/400297044-a82c4a16-276b-47d9-9b22-14ed8560753e.png)

**Comment by aivolcano** [comment]
Thanks for updates. Are you testing p-cot format data set. The final metric score is about 75.28, which outperforms SFT and online self training a lot in the paper. If we take the computational efficiency into account, I think 15 epochs are enough. Hahaha![image](https://user-images.githubusercontent.com/68730894/400298476-ca95ace8-d104-4eba-8645-0d37921f6d47.png)

**Comment by Pang-0093** [comment]
Yeah, python format. Yes, the performance improves rapidly in the initial ~15 epochs.

**Comment by aivolcano** [comment]
Good news!!!! what is the max epochs you run so far?  Is the score of 71.3%  the Value accuracy? 
<img width="1118" alt="image" src="https://github.com/user-attachments/assets/e6bf847e-44d8-4c1d-8c31-28108de83fa4" />

**Comment by lqtrung1998** [comment]
> Hi, I find that the RL stage runs so slowly. Each epoch consumes nearly two hours, given the bs=1 and 8 A100. The default setting in the paper suggests running RL for 300 epochs, which means that we have to complete the RL stage with almost 1 month 😢. Is there anything wrong with my situation? Or is it normal? If it is normal, can we do anything to accelerate the RL stage's training?

Hi @Pang-0093 ,
I believe 8A100 can fit `batch_size=4` as we also use 8A100 for training, you can try using the default setting. It is also suggested to increase `eval_batch_size` as much as possible as it will speed up evaluation (which is also time consuming to do evaluation at every epoch). If not working, I suggest keeping `batch_size=4` and using a tighter `max_gen_length` so that less memory is consumed.

For the expected training time, depending on the dataset, I believe it will not take longer than 2-3 weeks.

Best Regards,
Trung

**Comment by Pang-0093** [comment]
> Good news!!!! what is the max epochs you run so far? Is the score of 71.3% the Value accuracy? <img alt="image" width="1118" src="https://private-user-images.githubusercontent.com/68730894/400367487-e6bf847e-44d8-4c1d-8c31-28108de83fa4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzMzA2ODQsIm5iZiI6MTczNjMzMDM4NCwicGF0aCI6Ii82ODczMDg5NC80MDAzNjc0ODctZTZiZjg0N2UtNDRkOC00YzFkLThjMzEtMjgxMDhkZTgzZmE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDA5NTk0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFkODdlMzFlYWYwOGU4MWY1ZGNjZWNiOTRhN2Q1ZGExMTM5MWE1NDA1ZGJjNzZjNjQwZTM0ODRhNGQyYjc3MWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VZKFWjl5hJxbtjpFVlwS4EzwbxUWsIdsxJjsjsYtfFs">

Aha, I didn't run more epochs since I encountered a NCCL error and traced back. Yes, I mean the value accuracy, corresponding to the 75.8% in the table reported.

**Comment by Pang-0093** [comment]
> > Hi, I find that the RL stage runs so slowly. Each epoch consumes nearly two hours, given the bs=1 and 8 A100. The default setting in the paper suggests running RL for 300 epochs, which means that we have to complete the RL stage with almost 1 month 😢. Is there anything wrong with my situation? Or is it normal? If it is normal, can we do anything to accelerate the RL stage's training?
> 
> Hi @Pang-0093 , I believe 8A100 can fit `batch_size=4` as we also use 8A100 for training, you can try using the default setting. It is also suggested to increase `eval_batch_size` as much as possible as it will speed up evaluation (which is also time consuming to do evaluation at every epoch). If not working, I suggest keeping `batch_size=4` and using a tighter `max_gen_length` so that less memory is consumed.
> 
> For the expected training time, depending on the dataset, I believe it will not take longer than 2-3 weeks.
> 
> Best Regards, Trung

Thanks for your reply! I want to ensure that the 'batch_size' you mentioned represents the batch_size per GPU, right? Corresponds to the 'batch_size="4",mini_batch_size="4"' in _template.sh?

**Comment by aivolcano** [comment]
@Pang-0093 i think you can employ the flash attention to speed up the training process. The open source code is not calling the flash attention api.
